packages/babel-types/src/modifications/typescript/removeTypeDuplicates.js

```diff
@@ -0,0 +1,89 @@
+import {
+  isTSAnyKeyword,
+  isTSUnionType,
+  isTSBooleanKeyword,
+  isTSBigIntKeyword,
+  isTSNeverKeyword,
+  isTSNullKeyword,
+  isTSNumberKeyword,
+  isTSObjectKeyword,
+  isTSStringKeyword,
+  isTSUnknownKeyword,
+  isTSVoidKeyword,
+  isTSUndefinedKeyword,
+  isTSLiteralType,
+  isTSThisType,
+} from "../../validators/generated";
+
+/**
+ * Dedupe type annotations.
+ */
+export default function removeTypeDuplicates(
+  nodes: Array<Object>,
+): Array<Object> {
+  const generics = {};
+  const bases = {};
+
+  // store union type groups to circular references
+  const typeGroups = [];
+
+  const types = [];
+
+  for (let i = 0; i < nodes.length; i++) {
+    const node = nodes[i];
+    if (!node) continue;
+
+    // detect duplicates
+    if (types.indexOf(node) >= 0) {
+      continue;
+    }
+
+    // this type matches anything
+    if (isTSAnyKeyword(node.type)) {
+      return [node];
+    }
+
+    // Basic TS types, analogue of FlowBaseAnnotation
+    if (
+      isTSBooleanKeyword(node) ||
+      isTSBigIntKeyword(node) ||
+      isTSNeverKeyword(node) ||
+      isTSNullKeyword(node) ||
+      isTSNumberKeyword(node) ||
+      isTSObjectKeyword(node) ||
+      isTSStringKeyword(node) ||
+      isTSUndefinedKeyword(node) ||
+      isTSUnknownKeyword(node) ||
+      isTSVoidKeyword(node) ||
+      isTSThisType(node) ||
+      isTSLiteralType(node)
+    ) {
+      bases[node.type] = node;
+      continue;
+    }
+
+    if (isTSUnionType(node)) {
+      if (typeGroups.indexOf(node.types) < 0) {
+        nodes = nodes.concat(node.types);
+        typeGroups.push(node.types);
+      }
+      continue;
+    }
+
+    // TODO: add generic types
```

Comment:
⚠️ I need clarification here

First I tried `TSTypeParameter` https://github.com/babel/babel/pull/11378/commits/e0f3fc2dfd3d7320887e2259757afc03fce81271#diff-31b76153974cd0302007f05a570dfde6R49

Now I'm sure it does NOT cover all cases